src/mpl2/src/hier_rtlmp.cpp

```diff
@@ -4463,186 +4463,252 @@ bool Pusher::overlapsWithIOBlockage(const odb::Rect& cluster_box,
-  } else {
-    offset = layer->getOffsetY();
+      = getPinToLowerLeftDistance(pin, horizontal_snap);
+
+  // The distance between the pins and the lower-left corner
+  // of the master of a macro instance may not be a multiple
+  // of the track-grid, in these cases, we need to compensate
+  // a small offset.
+  int mterm_offset = 0;
+  if (params.pitch != 0) {
+    mterm_offset = params.lower_left_to_first_pin
+                   - std::floor(params.lower_left_to_first_pin / params.pitch)
+                         * params.pitch;
+  }
+  params.pin_offset = params.pin_width / 2 + mterm_offset;
+
+  const odb::dbOrientType& orientation = inst_->getOrient();
+  if (horizontal_snap) {
+    if (orientation == odb::dbOrientType::MY
+        || orientation == odb::dbOrientType::R180) {
+      params.pin_offset = -params.pin_offset;
+    }
+  } else if (orientation == odb::dbOrientType::MX
+             || orientation == odb::dbOrientType::R180) {
+    params.pin_offset = -params.pin_offset;
   }
-  return offset;
+  return params;
 }
 
-int Snapper::getPinWidth(odb::dbBox* box, const bool vertical_layer)
+int Snapper::getPinWidth(odb::dbITerm* pin, const bool horizontal_snap)
 {
   int pin_width = 0;
-  if (vertical_layer) {
-    pin_width = box->getDX();
+  if (horizontal_snap) {
+    pin_width = pin->getBBox().dx();
   } else {
-    pin_width = box->getDY();
+    pin_width = pin->getBBox().dy();
   }
   return pin_width;
 }
 
 void Snapper::getTrackGrid(odb::dbTrackGrid* track_grid,
                            std::vector<int>& coordinate_grid,
-                           const bool vertical_layer)
+                           const bool horizontal_snap)
 {
-  if (vertical_layer) {
+  if (horizontal_snap) {
     track_grid->getGridX(coordinate_grid);
   } else {
     track_grid->getGridY(coordinate_grid);
   }
 }
 
-int Snapper::getPinToLowerLeftDistance(odb::dbBox* box, bool vertical_layer)
+int Snapper::getPinToLowerLeftDistance(odb::dbITerm* pin, bool horizontal_snap)
 {
   int pin_to_origin = 0;
-  if (vertical_layer) {
-    pin_to_origin = box->getBox().xMin();
+
+  odb::dbMTerm* mterm = pin->getMTerm();
+  if (horizontal_snap) {
+    pin_to_origin = mterm->getBBox().xMin();
   } else {
-    pin_to_origin = box->getBox().yMin();
+    pin_to_origin = mterm->getBBox().yMin();
   }
+
   return pin_to_origin;
 }
 
+void Snapper::attemptSnapToExtraLayers(
+    const int origin,
+    const SameDirectionLayersData& layers_data,
+    const LayerParameters& snap_layer_params,
+    const bool horizontal_snap)
+{
+  const int total_number_of_layers
+      = static_cast<int>(layers_data.layer_to_pin.size());
+  const int total_attempts = 5;
+  int remaining_attempts = total_attempts;
+
+  int best_origin = origin;
+  int best_number_of_snapped_layers = 1;
+  int new_origin = origin;
+  bool all_layers_snapped = false;
+
+  while (remaining_attempts > 0) {
+    // Move one track ahead.
+    new_origin += snap_layer_params.pitch;
```

Comment:
Here you only move the origin considering the pitch of the same layer in every iteration?

Reply:
Yes, the snap_layer is the layer that we "manually" snapped. So we search for alignment in other layers by jumping its tracks.